addons/mail/static/src/model/model_core.js

```diff
@@ -2,134 +2,126 @@
+    }
+    if (!definition.modelMethods) {
+        definition.modelMethods = {};
+    }
+    for (const [methodName, method] of Object.entries(methods)) {
+        if (typeof method !== 'function') {
+            throw new Error(`Cannot patch model method on model "${modelName}": ${methodName} is not a function.`);
+        }
+        if (!definition.modelMethods[methodName]) {
+            definition.modelMethods[methodName] = method;
+            continue;
+        }
+        const methodBeforePatch = definition.modelMethods[methodName];
+        definition.modelMethods[methodName] = function(...args) {
+            this._super = methodBeforePatch;
+            return method.call(this, ...args);
+        };
+    }
 }
 
 /**
- * Register a patch for instance methods in model.
+ * Adds or overrides the provided methods to the records of the model specified
+ * by the `modelName`.
  *
- * @param {string} modelName
- * @param {string} patchName
- * @param {Object} patch
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {Object} methods
  */
-function registerInstancePatchModel(modelName, patchName, patch) {
-    _registerPatchModel(modelName, patchName, patch, { type: 'instance' });
+export function patchRecordMethods(modelName, methods) {
+    const definition = registry[modelName];
+    if (!definition) {
+        throw new Error(`Model "${modelName}" must be registered before patched.`);
+    }
+    if (!definition.recordMethods) {
+        definition.recordMethods = {};
+    }
+    for (const [methodName, method] of Object.entries(methods)) {
+        if (typeof method !== 'function') {
+            throw new Error(`Cannot patch record method on model "${modelName}": ${methodName} is not a function.`);
+        }
+        if (!definition.recordMethods[methodName]) {
+            definition.recordMethods[methodName] = method;
+            continue;
+        }
+        const methodBeforePatch = definition.recordMethods[methodName];
+        definition.recordMethods[methodName] = function(...args) {
+            this._super = methodBeforePatch;
+            return method.call(this, ...args);
+        };
+    }
 }
 
 /**
- * @param {string} name
- * @param {function} factory
- * @param {string[]} [dependencies=[]]
+ * @param {Object} definition The definition of the model to register.
  */
-function registerNewModel(name, factory, dependencies = []) {
-    const entry = _getEntryFromModelName(name);
-    let entryDependencies = [...dependencies];
-    if (name !== 'mail.model') {
-        entryDependencies = [...new Set(entryDependencies.concat(['mail.model']))];
+export function registerModel(definition) {
+    if (!definition.name) {
+        throw new Error("Model is lacking a name.");
+    }
+    if (!definition.identifyingFields) {
+        throw new Error(`Model "${definition.name}" is lacking identifying fields.`);
+    }
+    if (!definition.fields) {
+        definition.fields = {};
```

Comment:
Why are fields optionally added here, but methods are optionally added elsewhere?

Reply:
It would be better for consistency to do it here for all of them?